File: content/tracks/dsa-python/concepts/09-tree-dfs/tree-dfs_test.py
import sys
from stub import TreeNode, max_depth, has_path_sum, lowest_common_ancestor, _build

passed = 0
failed = 0


def test(name: str, condition: bool, detail: str = "") -> None:
    global passed, failed
    if condition:
        print(f"  PASS: {name}")
        passed += 1
    else:
        msg = f" — {detail}" if detail else ""
        print(f"  FAIL: {name}{msg}")
        failed += 1


def find_node(root: "TreeNode | None", val: int) -> "TreeNode | None":
    if root is None:
        return None
    if root.val == val:
        return root
    return find_node(root.left, val) or find_node(root.right, val)


# max_depth
test("depth: balanced h=3", max_depth(_build([3, 9, 20, None, None, 15, 7])) == 3)
test("depth: empty tree", max_depth(None) == 0)
test("depth: single node", max_depth(TreeNode(1)) == 1)
test("depth: two nodes", max_depth(_build([1, 2])) == 2)
test("depth: full 3 levels", max_depth(_build([1, 2, 3, 4, 5, 6, 7])) == 3)

# has_path_sum
test("pathsum: 5->4->11->2=22 exists", has_path_sum(_build([5, 4, 8, 11, None, 13, 4, 7, 2]), 22) is True)
test("pathsum: no path to 5", has_path_sum(_build([1, 2, 3]), 5) is False)
test("pathsum: empty tree", has_path_sum(None, 0) is False)
test("pathsum: single node matches", has_path_sum(TreeNode(7), 7) is True)
test("pathsum: single node no match", has_path_sum(TreeNode(7), 1) is False)
test("pathsum: negative values", has_path_sum(_build([-3, 1, -2]), -5) is True)

# lowest_common_ancestor
# Tree: [3,5,1,6,2,0,8,null,null,7,4]
root = _build([3, 5, 1, 6, 2, 0, 8, None, None, 7, 4])
p5 = find_node(root, 5)
p1 = find_node(root, 1)
p4 = find_node(root, 4)
p6 = find_node(root, 6)
p3 = find_node(root, 3)

lca = lowest_common_ancestor(root, p5, p1)
test("lca: 5,1 -> 3", lca is not None and lca.val == 3, f"got {lca.val if lca else None}")

lca = lowest_common_ancestor(root, p5, p4)
test("lca: 5,4 -> 5 (5 is ancestor)", lca is not None and lca.val == 5, f"got {lca.val if lca else None}")

lca = lowest_common_ancestor(root, p6, p4)
test("lca: 6,4 -> 5", lca is not None and lca.val == 5, f"got {lca.val if lca else None}")

lca = lowest_common_ancestor(root, p3, p1)
test("lca: root,1 -> root", lca is not None and lca.val == 3, f"got {lca.val if lca else None}")

print(f"\n{passed} passed, {failed} failed")
if failed > 0:
    sys.exit(1)

File: content/tracks/dsa-typescript/concepts/09-tree-dfs/tree-dfs_test.ts
import { TreeNode, maxDepth, hasPathSum, lowestCommonAncestor } from "./stub";

let passed = 0;
let failed = 0;

function test(name: string, fn: () => void): void {
  try {
    fn();
    console.log(`  PASS: ${name}`);
    passed++;
  } catch (e: unknown) {
    const msg = e instanceof Error ? e.message : String(e);
    console.log(`  FAIL: ${name} — ${msg}`);
    failed++;
  }
}

function assertEqual<T>(actual: T, expected: T): void {
  const a = JSON.stringify(actual);
  const b = JSON.stringify(expected);
  if (a !== b) throw new Error(`expected ${b}, got ${a}`);
}

// Build a tree from level-order array (null = missing node)
function buildTree(vals: (number | null)[]): TreeNode | null {
  if (vals.length === 0 || vals[0] === null) return null;
  const root = new TreeNode(vals[0]);
  const queue: TreeNode[] = [root];
  let i = 1;
  while (queue.length > 0 && i < vals.length) {
    const node = queue.shift()!;
    if (i < vals.length && vals[i] !== null) {
      node.left = new TreeNode(vals[i] as number);
      queue.push(node.left);
    }
    i++;
    if (i < vals.length && vals[i] !== null) {
      node.right = new TreeNode(vals[i] as number);
      queue.push(node.right);
    }
    i++;
  }
  return root;
}

// --- maxDepth ---

test("maxDepth: depth-3 tree", () => {
  assertEqual(maxDepth(buildTree([3, 9, 20, null, null, 15, 7])), 3);
});

test("maxDepth: null tree", () => {
  assertEqual(maxDepth(null), 0);
});

test("maxDepth: single node", () => {
  assertEqual(maxDepth(new TreeNode(1)), 1);
});

test("maxDepth: left-skewed", () => {
  const t = new TreeNode(1, new TreeNode(2, new TreeNode(3)));
  assertEqual(maxDepth(t), 3);
});

test("maxDepth: balanced depth 2", () => {
  assertEqual(maxDepth(buildTree([1, 2, 3])), 2);
});

// --- hasPathSum ---

test("hasPathSum: path exists (5→4→11→2=22)", () => {
  const tree = buildTree([5, 4, 8, 11, null, 13, 4, 7, 2, null, null, null, 1]);
  assertEqual(hasPathSum(tree, 22), true);
});

test("hasPathSum: path does not exist", () => {
  assertEqual(hasPathSum(buildTree([1, 2, 3]), 5), false);
});

test("hasPathSum: null tree", () => {
  assertEqual(hasPathSum(null, 0), false);
});

test("hasPathSum: single node match", () => {
  assertEqual(hasPathSum(new TreeNode(7), 7), true);
});

test("hasPathSum: single node no match", () => {
  assertEqual(hasPathSum(new TreeNode(7), 5), false);
});

test("hasPathSum: target matches internal node but not leaf path", () => {
  // Tree: root=1, left=2, right=3; path to left leaf = 3, path to right leaf = 4
  const t = buildTree([1, 2, 3]);
  // 1 itself is not a leaf; target=1 should not match
  assertEqual(hasPathSum(t, 1), false);
});

// --- lowestCommonAncestor ---

test("lowestCommonAncestor: p and q in different subtrees", () => {
  // Tree:    3
  //        /   \
  //       5     1
  //      / \   / \
  //     6   2 0   8
  //        / \
  //       7   4
  const root = buildTree([3, 5, 1, 6, 2, 0, 8, null, null, 7, 4]);
  // Build node references manually to use with LCA
  const p = root!.left!;      // node 5
  const q = root!.right!;     // node 1
  const lca = lowestCommonAncestor(root, p, q);
  assertEqual(lca?.val ?? null, 3);
});

test("lowestCommonAncestor: one is ancestor of the other", () => {
  const root = buildTree([3, 5, 1, 6, 2, 0, 8, null, null, 7, 4]);
  const p = root!.left!;              // node 5
  const q = root!.left!.right!.right!; // node 4
  const lca = lowestCommonAncestor(root, p, q);
  assertEqual(lca?.val ?? null, 5);
});

test("lowestCommonAncestor: p and q are both root's children", () => {
  const root = new TreeNode(1, new TreeNode(2), new TreeNode(3));
  const p = root.left!;
  const q = root.right!;
  assertEqual(lowestCommonAncestor(root, p, q)?.val, 1);
});

test("lowestCommonAncestor: single-node tree, p === q === root", () => {
  const root = new TreeNode(1);
  assertEqual(lowestCommonAncestor(root, root, root)?.val, 1);
});

console.log(`\n${passed} passed, ${failed} failed`);
if (failed > 0) process.exit(1);
